### DISCUSSION TITLE
understanding mqtt_as configuration process

### DISCUSSION THREAD

**Discussion by jfleon** [micropython/micropython#9709] [Using MicroPython] [answered]
Hi
I am using mqtt_as and it works beautifully our of the box.

when I changed some of its behavior to better suits my need I moved some of the built in value of the config dict in mqtt_as.py to the config.py file in a similar way than ssid and password are handled in the original library.
Then I realized the following
1 it works
2 I did not understand why  
indeed the only import of configuration I can find is 
`from mqtt_as import config`
so at this point in the program ssid and password  along with all th eparameters I dont want to freeze in mqtt_as are set to None
 I cannot figure out when the config.py file is read to update these values.

Can someone be kind enough to  explain to me when and how the file config.py is read to update the corresponding value in the config dict?
Thanks a lot

**Comment by beetlegigg** [comment]
In the config.py module the config dictionary from mqtt_as is imported and values are assigned to some dictionary items.  In the example config.py settings are made for the server the SSID and the wifi pw.   I understand from your post you have expanded this to include some other config settings in this config.py file.   In your main program file you must have imported the config.py file to enable all the settings in the config.py to take effect.   Taking the example file clean.py as an example you can see the line:
from config import wifi_led, blue_led 
which will mean the config.py file is read, the configuration settings are made, and the variables wifi_led and blue_led can be used in the clean.py module.   Does this answer your question ?

**Reply by davefes** [reply]
> In the config.py module the config dictionary from mqtt_as is imported ...

I see at the top of config.py this import statement but why does it appear to have to import itself?  Is there something special about the dictionary?

I read here [https://python-forum.io/thread-33092.html](url)the suggestion to put a dictionary in its own file, but that doesn't appear to be happening here.

Just interested in the process, a keyword(s) will do.

Thanks.

**Comment by jfleon** [comment]
Thanks guys
after reading your answers and experimenting a bit I think I understand how it works now.
the dictionary config is initially imported by the statement
`from mqtt_as import config`
and then in the statement
`from config import blue_led'`
the module config.py is read  blue_led is imported and in this process some previously imported value in the config dictionary are updated with the value they have in config.py : ssid and passwd in the published library.
I will have done it more explicit but it works.
I wonder if its a normal coding practice or just a trick used by the developers of the library for reasons of their own.

**Comment by beetlegigg** [comment, marked as answer]
Again taking the clean.py example for using the mqtt_as library, the  'from mqtt_as import config' in this file is only necessary as the clean.py code makes assignments to the config dictionary itself, and not all assignments are made just in the config.py file (the config.py file also has 'from mqtt_as import config' so likewise it can make changes to the config dictionary.
Too many things named config in these examples I must agree and in my use of this very useful mqtt_as library I do use a config file but I name my file mqtt_as_ini.py instead of config.py so I don't get confused.

**Reply by jfleon** [reply]
Yes and thank you
I had never encountered this before and it was counter intuitive to me at first- my ignorance and my bad-
.. and I agree with you about the config being everywhere.

**Reply by davefes** [reply]
Thanks for further explanation.  My stubbling-block is I understand that config.py would need "from mqtt_as import config" if the config dictionary does not reside in the config,py file.  Does the config dictionary start life in another file?

If the config dictionary starts life in config.py I will need to go away and try to understand why config.py couldn't modify it  ... without adding the "from mqtt_as import config".

**Reply by beetlegigg** [reply]
@davefes  - the config dictionary is created in the mqtt_as.py library and hence to assign or change the values in the dictionary in another python script, such as config.py then it needs to be imported into that file.   It maybe that you are seeing and using the mqtt_as.mpy file which is the pre-compiled file and not readable as text.  If you examine the mqtt_as.py file you can see the config dictionary being assigned.

**Reply by davefes** [reply]
Now I see it ... line 53 in mqtt_as.py.  Thank you for persisting.

**Comment by peterhinch** [comment]
I accept I made a mistake in naming the configuration file `config.py`, when the configuration dictionary is also called `config`. A clue is that each time I use it I have to remind myself how it works...

What is going on is as follows. A typical application starts with
```python
from mqtt_as import MQTTClient, config
```
This instantiates the `config` dictionary with default values. The key point of the next line
```python
from config import wifi_led, blue_led
```
is that `config.py` updates the `config` dictionary with local values such as WiFi credentials. This allows you to write your own `config.py` which you might copy to multiple devices or use with more than one application. I agree with @beetlegigg that it's a good idea to give `config.py` a different name.

The application then adds its own configuration details with lines like
```python
config['subs_cb'] = sub_cb
config['wifi_coro'] = wifi_han
config['will'] = (TOPIC, 'Goodbye cruel world!', False, 0)
```
before instantiating the client with
```python
client = MQTTClient(config)
```
Perhaps I should clarify this in the docs.
